materialize-bigquery/binding.go

```diff
@@ -1,225 +1,369 @@
+	query := client.Query(q)
+	query.TableDefinitions = map[string]bigquery.ExternalData{b.externalTableAlias: &b.writer.edc}
+	query.DefaultDatasetID = b.dataset
+	query.Location = b.region
+
+	return query.Run(ctx)
+}
+
+// Once Commit returns, the binding cannot be written to until Reset() is called.
+func (b *Binding) Commit(ctx context.Context) error {
+	return b.writer.commit(ctx)
+}
+
+// Reset is called when a new Writer need to be instantiated.
+// This is done so every pipeline operation in the bigquery materialization
+// can work with it's own writer to avoid data integrity issues.
+// Call Reset when the lifecycle of a bigquery materialization is completed and
+// the data is committed and acknowledged.
+func (b *Binding) Reset(ctx context.Context, name string) error {
+	var err error
+
+	if b.writer != nil {
+		err = b.writer.Destroy(ctx)
+		if err != nil {
+			return err
+		}
 	}
 
-	// External Data Config used by this binding. This will be a Google Cloud Storage bucket where we will
-	// create a file and then directly reference it as a table in a transaction.
-	b.load.extDataConfig = &bigquery.ExternalDataConfig{
+	obj := b.bucket.Object(name)
+	edc := bigquery.ExternalDataConfig{
 		SourceFormat: bigquery.JSON,
-		Schema:       make([]*bigquery.FieldSchema, 0),
+		Schema:       b.schema,
+		SourceURIs:   []string{fmt.Sprintf("gs://%s/%s", obj.BucketName(), obj.ObjectName())},
 	}
 
-	var pkJoins []string // How to join the main table to the external table based on keys.
+	b.writer, err = newWriter(ctx, edc, obj)
 
-	// We use the sqlDriver.NullableTypeMapping to handle null fields for defining schemas
-	// but when using bigquery it does not allow the NOT NULL suffix for external table
-	// columns. We need to get a reference to the base type mapper to leverage it's types.
-	baseTypeMapper := generator.TypeMappings.(sqlDriver.NullableTypeMapping).Inner
+	return err
+}
 
-	// Build the query for loading values from bigquery by joining a google cloud storage file
-	// based table of primary keys against the database.
-	for _, key := range spec.FieldSelection.Keys {
-		var col = tableDef.GetColumn(key)
+func (b *Binding) Store(doc map[string]interface{}) error {
+	err := b.writer.encoder.Encode(doc)
 
-		colType, err := baseTypeMapper.GetColumnType(col)
-		if err != nil {
-			return nil, err
-		}
+	return err
+}
+
+func (b *Binding) DeltaUpdates() bool {
+	return b.spec.DeltaUpdates
+}
 
-		b.load.extDataConfig.Schema = append(b.load.extDataConfig.Schema, &bigquery.FieldSchema{
-			Name:     identifierSanitizer(col.Name), // Sanitized name without quoting required
-			Repeated: false,
-			Required: col.NotNull,
-			Type:     bigquery.FieldType(colType.SQLType),
-		})
+// Whether the underlying writter has data committed
+// to the storage.
+func (b *Binding) DataCommitted() bool {
+	return b.writer.dataCommitted && b.writer.dataWritten > 0
```

Comment:
This is related to the big comment I left above, it's sort of a protection to skip any binding that doesn't have file to it. Maybe the name of the method/fields should be changed? I'll wait for us to discuss the above comment, because if my assumption around the transactor is proven wrong, this might not be necessary.